React/Modules/RCTDevSettings.m

```diff
@@ -0,0 +1,509 @@
+
+NSString *const kRCTDevSettingsUserDefaultsKey = @"RCTDevMenu";
+
+#if RCT_DEV
+
+@interface RCTDevSettingsUserDefaultsDataSource : NSObject <RCTDevSettingsDataSource>
+
+@property (nonatomic, strong) NSMutableDictionary *settings;
+@property (nonatomic, strong) NSMutableSet *keysChanged;
+
+@end
+
+@implementation RCTDevSettingsUserDefaultsDataSource
+
+- (instancetype)init
+{
+  return [self initWithDefaultValues:nil];
+}
+
+- (instancetype)initWithDefaultValues:(NSDictionary *)defaultValues
+{
+  if (self = [super init]) {
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(_userDefaultsDidChange)
+                                                 name:NSUserDefaultsDidChangeNotification
+                                               object:nil];
+    if (defaultValues) {
+      [self _reloadWithDefaults:defaultValues];
+    }
+    [self _userDefaultsDidChange];
+  }
+  return self;
+}
+
+- (void)dealloc
+{
+  [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)updateSettingWithValue:(id)value forKey:(NSString *)key
+{
+  RCTAssert((key != nil), @"%@", [NSString stringWithFormat:@"%@: Tried to update nil key", [self class]]);
+
+  id currentValue = _settings[key];
+  if (currentValue == value || [currentValue isEqual:value]) {
+    return;
+  }
+  if (value) {
+    _settings[key] = value;
+  } else {
+    [_settings removeObjectForKey:key];
+  }
+  if (!_keysChanged) {
+    _keysChanged = [NSMutableSet set];
+  }
+  [_keysChanged addObject:key];
+  [[NSUserDefaults standardUserDefaults] setObject:_settings forKey:kRCTDevSettingsUserDefaultsKey];
+  [[NSUserDefaults standardUserDefaults] synchronize];
+}
+
+- (id)settingForKey:(NSString *)key
+{
+  return _settings[key];
+}
+
+- (void)_userDefaultsDidChange
+{
+  NSDictionary *userInfo;
+  if (_keysChanged) {
+    userInfo = @{ kRCTDevSettingsUpdatedSettingsKey: [_keysChanged allObjects] };
+    _keysChanged = nil;
+  }
+  _settings = [NSMutableDictionary dictionaryWithDictionary:[[NSUserDefaults standardUserDefaults] objectForKey:kRCTDevSettingsUserDefaultsKey]];
+  [[NSNotificationCenter defaultCenter] postNotificationName:kRCTDevSettingsDidUpdateNotification object:self userInfo:userInfo];
+}
+
+- (void)_reloadWithDefaults:(NSDictionary *)defaultValues
+{
+  NSDictionary *existingSettings = [[NSUserDefaults standardUserDefaults] objectForKey:kRCTDevSettingsUserDefaultsKey];
+  _settings = (existingSettings) ? [existingSettings mutableCopy] : [NSMutableDictionary dictionary];
+  for (NSString *key in [defaultValues keyEnumerator]) {
+    if (!_settings[key]) {
+      _settings[key] = defaultValues[key];
+    }
+  }
+  [[NSUserDefaults standardUserDefaults] setObject:_settings forKey:kRCTDevSettingsUserDefaultsKey];
+  [[NSUserDefaults standardUserDefaults] synchronize];
+}
+
+@end
+
+@interface RCTDevSettings () <RCTBridgeModule, RCTInvalidating, RCTWebSocketObserverDelegate>
```

Comment:
`RCTWebSocketObserverDelegate` sounds like that should be split off, like RCTDevMenu, to another consumer of RCTDevSettings